papers/pydra/paper.rst

```diff
@@ -0,0 +1,467 @@
+
+This is schematically represented in Fig. :ref:`ndspl3`
+
+.. figure:: nd_spl_3-crop.pdf
+   :figclass: h!
+   :scale: 75%
+
+   Diagram representing a Task with two input fields and an outer splitter. :label:`ndspl3`
+
+
+In addition to the splitting the input, *Pydra* supports grouping or combining the output together.
+Taking as an example the simple *Task* represented in Fig. :ref:`ndspl1`, in some application
+it could be useful to combine all the outputs at the end.
+In order to do it *Task* has to have so called *combiner* that could be set by calling ``combine`` method.
+This could be written as follow:
+
+
+.. math::
+   :type: eqnarray
+
+   \textcolor{red}{\mathnormal{S} = x} &:& x=[x_1, x_2, ..., x_n] \longmapsto x=x_1, x=x_2, ..., x=x_n \\
+   \textcolor{red}{\mathnormal{C} = x} &:& out(x_1), out(x_2), ...out(x_n) \longmapsto out=[out(x_1), out(x_2), ...out(x_n)]
+
+
+Again, in the situation where input has multiple fields, there are various way of combining the output.
+Taking as an example *Task* represented in Fig. :ref:`ndspl4`, it could be useful to combine all the outputs
+for one specific values of *x* and all the values of *y*.
+The combining operation could be written as follow:
+
+.. math::
+   :type: eqnarray
+
+   \textcolor{red}{\mathnormal{C} = y} &:& out(x_1, y1), out(x_1, y2), ...out(x_n, y_m) \\
+    &\longmapsto& [[out(x_1, y_1), ..., out(x_1, y_m)], \\
+    && ..., \\
+    && [out(x_n, y_1), ..., out(x_n, y_m)]]
+
+And is represented in Fig. :ref:`ndspl3comb1` (todo: should probably change a,b to x,y)
+
+
+.. figure:: nd_spl_3_comb1-crop.pdf
+   :figclass: h!
+   :scale: 75%
+
+   Diagram representing a Task with two input fields, an outer splitter and a combiner. :label:`ndspl3comb1`
+
+It is also possible to combine all values of *x* for a specifc values of *y*, but it is also possible
+to combine all the values by providing a list of fields to the combiner.
+When a full combiner is set, i.e. all fields are in the combiner, the output is a one dimensional list:
+
+.. math::
+   :type: eqnarray
+
+   \textcolor{red}{\mathnormal{C} = [x, y]} : out(x_1, y1), ...out(x_n, y_m)
+    \longmapsto [out(x_1, y_1), ..., out(x_n, y_m)]
+
+
+And is represented in Fig. :ref:`ndspl3comb3` (todo: should probably change a,b to x,y)
+
+
+.. figure:: nd_spl_3_comb3-crop.pdf
+   :figclass: h!
+   :scale: 75%
+
+   Diagram representing a Task with two input fields, an outer splitter and a full combiner. :label:`ndspl3comb3`
+
+
+These are the basic examples of *splitters* and *combiners*, but *Pydra* allows for mixing
+*splitters* and *combiners* on various level.
+They could be set on a single *Task* level, or on *Workflow* level.
+They could be also passed from one *Task* to the followings within a *Workflow*.
+
+
+Global Cache
+============
+
+
+
+Applications and Examples
+-------------------------
+
+Machine Learning: Model Comparison
+==================================
+
+
+.. code-block:: python
+
+  ml example TODO
+
```

Comment:
```python
import pydra
import typing as ty
import os


# 1 - CREATE PYDRA TASKS
@pydra.mark.task # Convert python function to a Pydra Task object
@pydra.mark.annotate({"return": {"X": ty.Any, "Y": ty.Any, "groups": ty.Any}})  # Annotate to create named outputs
def read_file(filename, target_inds, x_indices=None, group='groups'):
   # Read and transform a CSV file to sklearn inputs
   import pandas as pd
   data = pd.read_csv(filename)
   X = data.iloc[:, x_indices]
   Y = data[target_inds]
   if group in data.keys():
       groups = data[:, [group]]
   else:
       groups = list(range(X.shape[0]))
   return X.values, Y.values, groups


@pydra.mark.task  # Convert python function to a Pydra Task object
@pydra.mark.annotate({"return": {"splits": ty.Any, "split_indices": ty.Any}})  # Annotate to create named outputs
def gen_splits(n_splits, test_size, X, Y, groups=None, random_state=0):
   # Generate a set of train-test splits
   from sklearn.model_selection import GroupShuffleSplit
   gss = GroupShuffleSplit(n_splits=n_splits, test_size=test_size,
                           random_state=random_state)
   train_test_splits = list(gss.split(X, Y, groups=groups))
   split_indices = list(range(n_splits))
   return train_test_splits, split_indices


@pydra.mark.task
@pydra.mark.annotate({"return": {"auc": ty.Any}})
def train_test_kernel(X, y, train_test_split, split_index, clf_info, permute):
   # Train and test a classifier on actual or permuted labels
   from sklearn.preprocessing import StandardScaler
   from sklearn.pipeline import Pipeline
   from sklearn.metrics import roc_auc_score
   mod = __import__(clf_info[0], fromlist=[clf_info[1]])
   clf = getattr(mod, clf_info[1])(**clf_info[2])
   if len(clf_info) > 3: # Run a GridSearch when param_grid available
       from sklearn.model_selection import GridSearchCV
       clf = GridSearchCV(clf, param_grid=clf_info[3])
   train_index, test_index = train_test_split[split_index]
   pipe = Pipeline([('std', StandardScaler()), (clf_info[1], clf)])
   y = y.ravel()
   if permute: # Run a generic permutation to create a null model
       pipe.fit(X[train_index], y[np.random.permutation(train_index)])
   else:
       pipe.fit(X[train_index], y[train_index])
   auc = roc_auc_score(y[test_index], pipe.predict(X[test_index]))
   return auc




# 2 - Specify inputs

# Suppose 'iris.csv' is the result of
from sklearn import datasets
X, y = datasets.load_iris(return_X_y=True)
iris_data = np.hstack((X, y.reshape((-1,1))))
print(iris_data.sample(10))

clfs = [
 ('sklearn.ensemble', 'ExtraTreesClassifier',
        dict(n_estimators=100, class_weight='balanced')),
 ('sklearn.neural_network', 'MLPClassifier', 
        dict(alpha=1, max_iter=1000)),
 ('sklearn.neighbors', 'KNeighborsClassifier', dict(),
        [{'n_neighbors': [3, 5, 7, 9, 11, 13, 15, 17, 19],
          'weights': ['uniform','distance']}]),
 ('sklearn.tree', 'DecisionTreeClassifier', 
        dict(max_depth=5)),
 ('sklearn.ensemble', 'AdaBoostClassifier', dict())]
inputs = {"filename": os.path.abspath('breast_cancer.csv'),
         "x_indices": range(30), "target_inds": -1,
         "n_splits": 5, "test_size": 0.2,
         "clf_info": clfs,  "permute": [True, False]
         }
n_procs = 8 # for parallel processing
cache_dir = os.path.join(os.getcwd(), 'cache')
wf_cache_dir = os.path.join(os.getcwd(), 'cache-wf')



# 3 - Use Pydra in Workflow style:

# Encapsulate tasks in a Workflow reuse script output cache
wf = pydra.Workflow(name="ml_wf", input_spec=list(inputs.keys()),
               **inputs, cache_dir=wf_cache_dir, # workflow cache 
                   cache_locations=[cache_dir]) # reuses script cache

# joint map over classifiers and permutation
wf.split(['clf_info', 'permute']) 
wf.add(read_file(name="readcsv",               # add task
                filename=wf.lzin.filename,     # connect workflow input
                x_indices=wf.lzin.x_indices,
                target_inds=wf.lzin.target_inds))
wf.add(gen_splits(name="gensplit",             # add task 
                 n_splits=wf.lzin.n_splits,    # connect workflow input
                 test_size=wf.lzin.test_size,
                 # connect lazy-eval outputs of previous task
                 X=wf.readcsv.lzout.X, Y=wf.readcsv.lzout.Y,
                 groups=wf.readcsv.lzout.groups))
wf.add(train_test_kernel(name="fit_clf", # use outputs from both tasks
                   X=wf.readcsv.lzout.X, y=wf.readcsv.lzout.Y,
                   train_test_split=wf.gensplit.lzout.splits,
                   split_index=wf.gensplit.lzout.split_indices,
                   clf_info=wf.lzin.clf_info, permute=wf.lzin.permute))
wf.fit_clf.split('split_index').combine('split_index') # Parallel spec
wf.set_output([("auc", wf.fit_clf.lzout.auc)]) # connect workflow output

#Execute the workflow in parallel using multiple processes
with pydra.Submitter(plugin="cf", n_procs=n_procs) as sub:
   sub(runnable=wf)

print(wf.result(return_inputs=True))

```

Reply:
it would be great if you can work on this part, however if you work on big snippet of the code, you should work on this locally and build the pdf paper.  You will see that unfortunately the lines of code have to be short to fit nicely to the column.

Reply:
so this is why i do not merge it here